grub-core/commands/blscfg.c

```diff
@@ -1163,6 +1236,464 @@ bls_create_entries (bool show_default, bool show_non_default, char *entry_id)
+	{
+	  *e = '\0';
+	  --e;
+	}
+
+      /* strip closing " */
+      if (quotes)
+	{
+	  e--;
+	  if (e >= line && *e == '"')
+	    *e = '\0';
+	}
+
+    return GRUB_ERR_NONE;
+
+    }
+
+  return GRUB_ERR_EOF;
+}
+
+static int
+add_uki_bls_entry (const char *filename UNUSED, const char *path,
+		   const char *title)
+{
+  struct bls_entry *entry;
+  int status;
+
+  entry = grub_zalloc (sizeof (*entry));
+  if (entry == NULL)
+    return GRUB_ERR_OUT_OF_MEMORY;
+
+  // 0 means visible.
+  entry->visible = 0;
+
+  entry->filename = grub_strdup (path);
+  if (!entry->filename)
+    {
+      status = GRUB_ERR_OUT_OF_MEMORY;
+      goto cleanup;
+    }
+
+  status = bls_add_keyval (entry, "efi", path);
+  if (status)
+    goto cleanup;
+
+  status = bls_add_keyval (entry, "title", title);
+  if (status)
+    goto cleanup;
+
+  status = bls_add_entry (entry);
+  if (!status)
+    return GRUB_ERR_NONE;
+
+ cleanup:
+  grub_free (entry);
+  return status;
+}
+
+static int
+read_uki (const char *filename,
+	  const struct grub_dirhook_info *dirhook_info UNUSED,
+	  void *data)
+{
+  grub_size_t n;
+  char *path;
+  grub_file_t file;
+  grub_err_t status;
+  struct read_entry_info *info = (struct read_entry_info *)data;
+  struct grub_pe32_section_table osrel_section_header;
+  char *section_data = NULL;
+  char *title = NULL;
+
+  grub_dprintf ("blscfg", "efi filename: \"%s\"\n", filename);
+
+  n = grub_strlen (filename);
+
+  if (filename[0] == '.')
+    return 0;
+
+  if (n <= sizeof (".efi"))
+    return 0;
+
+  if (grub_strcmp (filename + n - 4, ".efi") != 0)
+    return 0;
+
+  path = grub_xasprintf ("(%s)%s/%s", info->devid, info->dirname, filename);
+  if (path == NULL)
+    return 0;
+
+  file = grub_file_open (path, GRUB_FILE_TYPE_NONE);
+  if (!file)
+    {
+      grub_dprintf ("blscfg", "Error opening file %s", filename);
+      return 0;
+    }
+
+  status = get_pe32_section_header (file, ".osrel", &osrel_section_header);
+  if (status != GRUB_ERR_NONE)
+    {
+      grub_dprintf ("blscfg", "Did not find '.osrel' section in efi file\n");
+      goto finish;
+    }
+
+  if (osrel_section_header.raw_data_size > OSREL_SECTION_SIZE_MAX)
+    {
+      grub_dprintf ("blscfg", "'.osrel' section too large: %d",
+		    osrel_section_header.raw_data_size);
+      goto finish;
+    }
+
+  section_data = grub_malloc (osrel_section_header.raw_data_size);
+  if (section_data == NULL)
+    {
+      grub_error (GRUB_ERR_OUT_OF_MEMORY, N_("Out of memory"));
+      goto finish;
+    }
+
+  grub_file_seek (file, osrel_section_header.raw_data_offset);
+  n = grub_file_read (file, section_data, osrel_section_header.raw_data_size);
+  if (n != osrel_section_header.raw_data_size)
+    {
+      grub_dprintf ("blscfg", "Error reading section data from file\n");
+      goto finish;
+    }
+
+  char *key;
+  char *value;
+  const char *pretty_name = "(no title)";
+  char *osrel = section_data;
+  while (osrel_get_key_value (&osrel, &key, &value) == GRUB_ERR_NONE)
```

Comment:
This looks like it'll search off the end if the data is malformed (i.e. has no newlines).

Reply:
The whole osrel data might not be null terminated, and then we might run off the end. No `\n` are not a problem.